File: Demo/Data/2_loc_auto_key/Output/Client/Cpp/table_loader.cpp

#include <iostream>
#include "table_loader.h"

namespace Test {
	const std::string TableLoaderMgr::LangList[] = {"zh-Hans","EN"};	

	template <class T1, class T2>
	static bool _ReadTuple(ITableRowReader* reader, std::tuple<T1, T2>& item)
	{
		auto tupleReader= reader->BeginTuple();
		if (tupleReader == nullptr)
			return true;
		tupleReader->Read(std::get<0>(item));
		tupleReader->Read(std::get<1>(item));
		return true;
	}


	template <class T1, class T2>
	static bool _ReadListTuple(ITableRowReader* reader, std::vector<std::tuple<T1, T2>>& list)
	{
		auto listReader = reader->BeginList();
		if (listReader == nullptr)
			return true;
		int count = listReader->GetCount();
		list.resize(count);
		list.clear();
		for (int i = 0; i < count; i++)
		{
			std::tuple<T1, T2> item;
			auto tupleReader = listReader->NextTuple();
			tupleReader->Read(std::get<0>(item));
			tupleReader->Read(std::get<1>(item));
			list.push_back(item);
		}
		return true;
	}

	template <class T1, class T2, class T3>
	static bool _ReadTuple(ITableRowReader* reader, std::tuple<T1, T2, T3>& item)
	{
		auto tupleReader = reader->BeginTuple();
		if (tupleReader == nullptr)
			return true;
		tupleReader->Read(std::get<0>(item));
		tupleReader->Read(std::get<1>(item));
		tupleReader->Read(std::get<2>(item));
		return true;
	}

	template <class T1, class T2, class T3>
	static bool _ReadListTuple(ITableRowReader* reader, std::vector<std::tuple<T1, T2, T3>>& list)
	{
		auto listReader = reader->BeginList();
		if (listReader == nullptr)
			return true;
		int count = listReader->GetCount();
		list.resize(count);
		list.clear();
		for (int i = 0; i < count; i++)
		{
			std::tuple<T1, T2, T3> item;
			auto tupleReader = listReader->NextTuple();
			tupleReader->Read(std::get<0>(item));
			tupleReader->Read(std::get<1>(item));
			tupleReader->Read(std::get<2>(item));
			list.push_back(item);
		}
		return true;
	}


	template <class T1, class T2, class T3, class T4>
	static bool _ReadTuple(ITableRowReader* reader, std::tuple<T1, T2, T3, T4>& item)
	{
		auto tupleReader = reader->BeginTuple();
		if (tupleReader == nullptr)
			return true;
		tupleReader->Read(std::get<0>(item));
		tupleReader->Read(std::get<1>(item));
		tupleReader->Read(std::get<2>(item));
		tupleReader->Read(std::get<3>(item));
		return true;
	}

	template <class T1, class T2, class T3, class T4>
	static bool _ReadListTuple(ITableRowReader* reader, std::vector<std::tuple<T1, T2, T3, T4>>& list)
	{
		auto listReader = reader->BeginList();
		if (listReader == nullptr)
			return true;
		int count = listReader->GetCount();
		list.resize(count);
		list.clear();
		for (int i = 0; i < count; i++)
		{
			std::tuple<T1, T2, T3, T4> item;
			auto tupleReader = listReader->NextTuple();
			tupleReader->Read(std::get<0>(item));
			tupleReader->Read(std::get<1>(item));
			tupleReader->Read(std::get<2>(item));
			tupleReader->Read(std::get<3>(item));
			list.push_back(item);
		}
		return true;
	}

	template <class T1, class T2, class T3, class T4, class T5>
	static bool _ReadTuple(ITableRowReader* reader, std::tuple<T1, T2, T3, T4, T5>& item)
	{
		auto tupleReader = reader->BeginTuple();
		if (tupleReader == nullptr)
			return true;
		tupleReader->Read(std::get<0>(item));
		tupleReader->Read(std::get<1>(item));
		tupleReader->Read(std::get<2>(item));
		tupleReader->Read(std::get<3>(item));
		tupleReader->Read(std::get<4>(item));
		return true;
	}

	template <class T1, class T2, class T3, class T4, class T5>
	static bool _ReadListTuple(ITableRowReader* reader, std::vector<std::tuple<T1, T2, T3, T4, T5>>& list)
	{
		auto listReader = reader->BeginList();
		if (listReader == nullptr)
			return true;
		int count = listReader->GetCount();
		list.resize(count);
		list.clear();
		for (int i = 0; i < count; i++)
		{
			std::tuple<T1, T2, T3, T4, T5> item;
			auto tupleReader = listReader->NextTuple();
			tupleReader->Read(std::get<0>(item));
			tupleReader->Read(std::get<1>(item));
			tupleReader->Read(std::get<2>(item));
			tupleReader->Read(std::get<3>(item));
			tupleReader->Read(std::get<4>(item));
			tupleReader.push_back(item);
		}
		return true;
	}


	template <class T>
	static bool _ReadEnum(ITableRowReader* reader, T& v)
	{
		int vv;
		if (!reader->Read(vv))
			return false;
		v = (T)vv;
		return true;
	}

	template <class T>
	static bool _ReadList(ITableRowReader* reader, std::vector<T>& list)
	{
		auto listReader = reader->BeginList();
		if (listReader == nullptr)
			return true;
		int count = listReader->GetCount();
		list.resize(count);
		list.clear();
		for (int i = 0; i < count; i++)
		{
			T item;
			listReader->Read(item);
			list.push_back(item);
		}
		return true;
	}


static Table* _LoadItemData(Table* table, std::string lang, ITableReaderCreator& tableReaderCreator)
{
    auto tableList = dynamic_cast<TableList<TItemData>*>(table);

    std::string sheet_name = "ItemData";
    lang = "";
    int col_count = 11;

    ITableReader* reader = tableReaderCreator.ReadTable(sheet_name,lang);
    if(reader == nullptr)
        return nullptr;

    //Check Header
    const std::vector<std::string>& header = reader->ReadHeader();
    if (header.size() != (col_count*2))
    {
		std::cerr << "加载错误,表头数量不对" << sheet_name << std::endl;
        return nullptr;
    }
    bool head_rst = true;
    
	head_rst &= (header[0].compare("Id")==0) && (header[0+11].compare("int32")==0);
	head_rst &= (header[1].compare("Name")==0) && (header[1+11].compare("locid")==0);
	head_rst &= (header[2].compare("Type")==0) && (header[2+11].compare("int32")==0);
	head_rst &= (header[3].compare("SubType")==0) && (header[3+11].compare("int32")==0);
	head_rst &= (header[4].compare("Quality")==0) && (header[4+11].compare("int32")==0);
	head_rst &= (header[5].compare("PairField")==0) && (header[5+11].compare("int32_bool")==0);
	head_rst &= (header[6].compare("PairField2")==0) && (header[6+11].compare("int32_bool")==0);
	head_rst &= (header[7].compare("PairField3")==0) && (header[7+11].compare("int32_int32")==0);
	head_rst &= (header[8].compare("PairFieldList")==0) && (header[8+11].compare("list_int32_int64")==0);
	head_rst &= (header[9].compare("PairFieldList2")==0) && (header[9+11].compare("list_int32_int64")==0);
	head_rst &= (header[10].compare("ListField")==0) && (header[10+11].compare("list_int32")==0);

    if (!head_rst)
    {
		std::cerr << "加载错误,表头不匹配" << sheet_name << std::endl;
        return nullptr;
    }

    int rowCount = reader->ReadRowCount();
    TItemData* array = new TItemData[rowCount];
    //加载数据
    for (int i = 0; i < rowCount; i++)
    { 
		auto rowReader = reader->NextRow();
        TItemData item;
		rowReader->Read(item.Id);
		rowReader->Read(item.Name);
		_ReadEnum(rowReader,item.Type);
		_ReadEnum(rowReader,item.SubType);
		_ReadEnum(rowReader,item.Quality);
		_ReadTuple(rowReader, item.PairField);
		_ReadTuple(rowReader, item.PairField2);
		_ReadTuple(rowReader, item.PairField3);
		_ReadListTuple(rowReader, item.PairFieldList);
		_ReadListTuple(rowReader, item.PairFieldList2);
		_ReadList(rowReader, item.ListField);

        array[i] = item;
    }
    tableList->SetListData(array, rowCount);  


		auto tableDict= dynamic_cast<TableDict<int,TItemData>*>(table);
		tableDict->Dict.clear();
		for (int i = 0; i < rowCount; i++)
		{
			TItemData* item = array + i;
			int key = item->Id;			 
			tableDict->Dict[key] = item;
			
		}
	return tableList;
}

static Table* _LoadTestComposeKey(Table* table, std::string lang, ITableReaderCreator& tableReaderCreator)
{
    auto tableList = dynamic_cast<TableList<TTestComposeKey>*>(table);

    std::string sheet_name = "TestComposeKey";
    lang = "";
    int col_count = 5;

    ITableReader* reader = tableReaderCreator.ReadTable(sheet_name,lang);
    if(reader == nullptr)
        return nullptr;

    //Check Header
    const std::vector<std::string>& header = reader->ReadHeader();
    if (header.size() != (col_count*2))
    {
		std::cerr << "加载错误,表头数量不对" << sheet_name << std::endl;
        return nullptr;
    }
    bool head_rst = true;
    
	head_rst &= (header[0].compare("Id")==0) && (header[0+5].compare("uint32")==0);
	head_rst &= (header[1].compare("Level")==0) && (header[1+5].compare("int32")==0);
	head_rst &= (header[2].compare("Name")==0) && (header[2+5].compare("locid")==0);
	head_rst &= (header[3].compare("Pos")==0) && (header[3+5].compare("float32_float32_float32")==0);
	head_rst &= (header[4].compare("Flags")==0) && (header[4+5].compare("int32")==0);

    if (!head_rst)
    {
		std::cerr << "加载错误,表头不匹配" << sheet_name << std::endl;
        return nullptr;
    }

    int rowCount = reader->ReadRowCount();
    TTestComposeKey* array = new TTestComposeKey[rowCount];
    //加载数据
    for (int i = 0; i < rowCount; i++)
    { 
		auto rowReader = reader->NextRow();
        TTestComposeKey item;
		rowReader->Read(item.Id);
		rowReader->Read(item.Level);
		rowReader->Read(item.Name);
		_ReadTuple(rowReader, item.Pos);
		_ReadEnum(rowReader,item.Flags);

        array[i] = item;
    }
    tableList->SetListData(array, rowCount);  


		auto tableDict= dynamic_cast<TableDict<std::tuple<unsigned int,int>,TTestComposeKey>*>(table);
		tableDict->Dict.clear();
		for (int i = 0; i < rowCount; i++)
		{
			TTestComposeKey* item = array + i;
			std::tuple<unsigned int,int> key(item->Id,item->Level);
			tableDict->Dict[key] = item;
			
		}
	return tableList;
}

static Table* _LoadLoc(Table* table, std::string lang, ITableReaderCreator& tableReaderCreator)
{
    auto tableList = dynamic_cast<TableList<TLoc>*>(table);

    std::string sheet_name = "Loc";
    
    int col_count = 2;

    ITableReader* reader = tableReaderCreator.ReadTable(sheet_name,lang);
    if(reader == nullptr)
        return nullptr;

    //Check Header
    const std::vector<std::string>& header = reader->ReadHeader();
    if (header.size() != (col_count*2))
    {
		std::cerr << "加载错误,表头数量不对" << sheet_name << std::endl;
        return nullptr;
    }
    bool head_rst = true;
    
	head_rst &= (header[0].compare("Id")==0) && (header[0+2].compare("int32")==0);
	head_rst &= (header[1].compare("Val")==0) && (header[1+2].compare("string")==0);

    if (!head_rst)
    {
		std::cerr << "加载错误,表头不匹配" << sheet_name << std::endl;
        return nullptr;
    }

    int rowCount = reader->ReadRowCount();
    TLoc* array = new TLoc[rowCount];
    //加载数据
    for (int i = 0; i < rowCount; i++)
    { 
		auto rowReader = reader->NextRow();
        TLoc item;
		rowReader->Read(item.Id);
		rowReader->Read(item.Val);

        array[i] = item;
    }
    tableList->SetListData(array, rowCount);  


		auto tableDict= dynamic_cast<TableDict<int,TLoc>*>(table);
		tableDict->Dict.clear();
		for (int i = 0; i < rowCount; i++)
		{
			TLoc* item = array + i;
			int key = item->Id;			 
			tableDict->Dict[key] = item;
			
		}
	return tableList;
}

TableLoaderMgr::TableLoaderMgr()
{
	LoaderDict["TItemData"] = TableInfo(_LoadItemData, false);
	LoaderDict["TTestComposeKey"] = TableInfo(_LoadTestComposeKey, false);
	LoaderDict["TLoc"] = TableInfo(_LoadLoc, true);
}
}

File: Demo/Data/2_loc_auto_key/Output/Server/CS/table_loader.cs
//自动生成的
using System;
using System.Collections;
using System.Collections.Generic;
namespace Test{

    public delegate bool CreateTableReader(string sheet_name, string lang_name, out ITableReader reader);
    public delegate Table TableLoader(string lang);
    public struct TableInfo
    {
        public TableLoader Loader;
        public bool MultiLang;
        public TableInfo(TableLoader loader, bool multiLang)
        {
            this.Loader = loader;
            this.MultiLang = multiLang;
        }
    }
    public partial class TableLoaderMgr
    {
        private static List<System.Object> _temp = new List<System.Object>(10000);
        public static List<string> LangList;

        public CreateTableReader CreateTableReader;
        public Dictionary<Type, TableInfo> LoaderDict;
        static TableLoaderMgr()
        {        
            LangList= new List<string>(2);

			LangList.Add("zh-Hans");
			LangList.Add("EN");
			EnumConverterMgr.RegFunc((v) => (EItemType)v, (v) => (int)v);
			EnumConverterMgr.RegFunc((v) => (EItemSubType)v, (v) => (int)v);
			EnumConverterMgr.RegFunc((v) => (EItemQuality)v, (v) => (int)v);
			EnumConverterMgr.RegFunc((v) => (EItemFlag)v, (v) => (int)v);
		}

        public TableLoaderMgr(CreateTableReader createTableReader)
        {
            CreateTableReader = createTableReader;            
            LoaderDict = new Dictionary<Type, TableInfo>(20+2);
            

			LoaderDict.Add(typeof(TItemData),new TableInfo(_LoadItemData,false));
			LoaderDict.Add(typeof(TTestComposeKey),new TableInfo(_LoadTestComposeKey,false));
		}

        private Table _LoadItemData(string lang)
        {
            string sheet_name = "ItemData";
            lang = null;
            int col_count = 10;

            if(!CreateTableReader(sheet_name,lang,out var reader))
                return null;

            //Check Header
            var header = reader.ReadHeader();
            if (header==null || header.Count != (col_count*2))
            {
                Log.E("加载错误 {0},表头数量不对", sheet_name);
                return null;
            }
            bool head_rst = true;
            
			head_rst &= ((header[0] == "Id") && (header[0+10] == "int32"));
			head_rst &= ((header[1] == "Type") && (header[1+10] == "int32"));
			head_rst &= ((header[2] == "SubType") && (header[2+10] == "int32"));
			head_rst &= ((header[3] == "Quality") && (header[3+10] == "int32"));
			head_rst &= ((header[4] == "PairField") && (header[4+10] == "int32_bool"));
			head_rst &= ((header[5] == "PairField2") && (header[5+10] == "int32_bool"));
			head_rst &= ((header[6] == "PairField3") && (header[6+10] == "int32_int32"));
			head_rst &= ((header[7] == "PairFieldList") && (header[7+10] == "list_int32_int64"));
			head_rst &= ((header[8] == "PairFieldList2") && (header[8+10] == "list_int32_int64"));
			head_rst &= ((header[9] == "ListField") && (header[9+10] == "list_int32"));

            if (!head_rst)
            {
                Log.E("加载错误 {0}, 表头不匹配", sheet_name);
                return null;
            }

            //加载数据
            _temp.Clear();
            for (; ; )
            {
                if (!reader.NextRow(out var rowReader))
                    break;                
                var row = new TItemData();
				_Read(rowReader, ref row.Id);
				_Read(rowReader, ref row.Type);
				_Read(rowReader, ref row.SubType);
				_Read(rowReader, ref row.Quality);
				_ReadTuple(rowReader.BeginTuple(), ref row.PairField);
				_ReadTuple(rowReader.BeginTuple(), ref row.PairField2, out (int,bool) __PairField2);
				_ReadTuple(rowReader.BeginTuple(), ref row.PairField3, out (int,int) __PairField3);
				_ReadList(rowReader, ref row.PairFieldList);
				_ReadList(rowReader, ref row.PairFieldList2,out (int,long) __PairFieldList2);
				_ReadList(rowReader, ref row.ListField);

                _temp.Add(row);
            }

            //转换数据
            List<TItemData> list = new List<TItemData>(_temp.Count);
            foreach (var p in _temp)
            {
                list.Add(p as TItemData);
            }            
            

            var dict = new Dictionary<int, TItemData>(list.Count);
            foreach (var p in list)
            {
                if (dict.ContainsKey(p.Id))
                {
                    Log.E("{0} Contain Multi Id: {1}, 如果允许ID重复, 修改表格", typeof(TItemData), p.Id);
                    continue;
                }
                dict.Add(p.Id, p);
            }
            return Table.Create(list,dict);
                
		}

        private Table _LoadTestComposeKey(string lang)
        {
            string sheet_name = "TestComposeKey";
            lang = null;
            int col_count = 5;

            if(!CreateTableReader(sheet_name,lang,out var reader))
                return null;

            //Check Header
            var header = reader.ReadHeader();
            if (header==null || header.Count != (col_count*2))
            {
                Log.E("加载错误 {0},表头数量不对", sheet_name);
                return null;
            }
            bool head_rst = true;
            
			head_rst &= ((header[0] == "Id") && (header[0+5] == "uint32"));
			head_rst &= ((header[1] == "Level") && (header[1+5] == "int32"));
			head_rst &= ((header[2] == "Name") && (header[2+5] == "locid"));
			head_rst &= ((header[3] == "Pos") && (header[3+5] == "float32_float32_float32"));
			head_rst &= ((header[4] == "Flags") && (header[4+5] == "int32"));

            if (!head_rst)
            {
                Log.E("加载错误 {0}, 表头不匹配", sheet_name);
                return null;
            }

            //加载数据
            _temp.Clear();
            for (; ; )
            {
                if (!reader.NextRow(out var rowReader))
                    break;                
                var row = new TTestComposeKey();
				_Read(rowReader, ref row.Id);
				_Read(rowReader, ref row.Level);
				_Read(rowReader, ref row.Name);
				_ReadTuple(rowReader.BeginTuple(), ref row.Pos);
				_Read(rowReader, ref row.Flags);

                _temp.Add(row);
            }

            //转换数据
            List<TTestComposeKey> list = new List<TTestComposeKey>(_temp.Count);
            foreach (var p in _temp)
            {
                list.Add(p as TTestComposeKey);
            }            
            

            var dict = new Dictionary<(uint,int), TTestComposeKey>(list.Count);
            foreach (var p in list)
            {
                (uint,int) key = (p.Id, p.Level);
                if (dict.ContainsKey(key))
                {
                    Log.E("{0} Contain Multi Id: {1},{2}, 如果允许ID重复, 修改表格", typeof(TTestComposeKey), p.Id,p.Level);
                    continue;
                }
                dict.Add(key, p);
            }
            return Table.Create(list,dict);
                
		}

        #region Base Reader
        private static void _Read(ITableDataReader reader, ref bool v)
        {
            v = reader.ReadBool();
        }
        private static void _Read(ITableDataReader reader, ref int v)
        {
            v = reader.ReadInt32();
        }
        private static void _Read(ITableDataReader reader, ref uint v)
        {
            v = reader.ReadUInt32();
        }
        private static void _Read(ITableDataReader reader, ref long v)
        {
            v = reader.ReadInt64();
        }
        private static void _Read(ITableDataReader reader, ref ulong v)
        {
            v = reader.ReadUInt64();
        }
        private static void _Read(ITableDataReader reader, ref float v)
        {
            v = reader.ReadF32();
        }
        private static void _Read(ITableDataReader reader, ref double v)
        {
            v = reader.ReadF64();
        }
        private static void _Read(ITableDataReader reader, ref string v)
        {
            v = reader.ReadString();
        }
        private static void _Read(ITableDataReader reader, ref LocStr v)
        {
            string s = reader.ReadString();
            v = s;
        }
        private static void _Read(ITableDataReader reader, ref LocId v)
        {
            v = reader.ReadInt32();
        }
        private static void _Read<T>(ITableDataReader reader, ref T v) where T : Enum
        {
            if (!EnumConverterMgr.Convert(reader.ReadInt32(), ref v))
            {
                Log.E("没有找到类型 {0} 的转换", typeof(T));
            }
        }
        #endregion
		#region Tuple Reader

        private static void _ReadTuple(ITableTupleReader tupleReader, ref (int,bool) v)
        {
            if(tupleReader==null)
                return;

			_Read(tupleReader,ref v.Item1);
			_Read(tupleReader,ref v.Item2);
		}

        private static void _ReadTuple(ITableTupleReader tupleReader, ref PairItemIntBool v, out (int,bool) v2)
        {
            v2=default;
            if(tupleReader==null)
            {
                v= PairItemIntBool.CreateInst(false,v2);
                return;
            }
			_Read(tupleReader,ref v2.Item1);
			_Read(tupleReader,ref v2.Item2);

             v = PairItemIntBool.CreateInst(true,v2);
        }

        private static void _ReadTuple(ITableTupleReader tupleReader, ref PairItemIntBool v, out (int,int) v2)
        {
            v2=default;
            if(tupleReader==null)
            {
                v= PairItemIntBool.CreateInst(false,v2);
                return;
            }
			_Read(tupleReader,ref v2.Item1);
			_Read(tupleReader,ref v2.Item2);

             v = PairItemIntBool.CreateInst(true,v2);
        }

        private static void _ReadTuple(ITableTupleReader tupleReader, ref (int,long) v)
        {
            if(tupleReader==null)
                return;

			_Read(tupleReader,ref v.Item1);
			_Read(tupleReader,ref v.Item2);
		}

        private static void _ReadTuple(ITableTupleReader tupleReader, ref PairItemIntInt64 v, out (int,long) v2)
        {
            v2=default;
            if(tupleReader==null)
            {
                v= PairItemIntInt64.CreateInst(false,v2);
                return;
            }
			_Read(tupleReader,ref v2.Item1);
			_Read(tupleReader,ref v2.Item2);

             v = PairItemIntInt64.CreateInst(true,v2);
        }

        private static void _ReadTuple(ITableTupleReader tupleReader, ref (float,float,float) v)
        {
            if(tupleReader==null)
                return;

			_Read(tupleReader,ref v.Item1);
			_Read(tupleReader,ref v.Item2);
			_Read(tupleReader,ref v.Item3);
		}
		#endregion

		#region List Reader

        private static void _ReadList(ITableRowReader rowReader, ref (int,long)[]v)
        {
            var listReader = rowReader.BeginList();
            int count = listReader != null ? listReader.GetCount() : 0;
            if (count == 0)
                v = Array.Empty<(int,long)>();
            else
            {
                v = new (int,long)[count];
                for (int i = 0; i < count; i++)
                {
                    (int,long) item = default;
                    _ReadTuple(listReader.BeginTuple(), ref item);                    
                    v[i] = item;
                }
            }
        }

        private static void _ReadList(ITableRowReader rowReader, ref PairItemIntInt64[] v, out (int,long) v2)
        {
            v2=default;
            var listReader = rowReader.BeginList();
            int count = listReader != null ? listReader.GetCount() : 0;
            if (count == 0)
                v = Array.Empty<PairItemIntInt64>();
            else
            {
                v = new PairItemIntInt64[count];
                for (int i = 0; i < count; i++)
                {
                    PairItemIntInt64 item = default;
                    _ReadTuple(listReader.BeginTuple(), ref item, out v2);
                    v[i] = item;
                }
            }
        }

        private static void _ReadList(ITableRowReader rowReader, ref int[]v)
        {
            var listReader = rowReader.BeginList();
            int count = listReader != null ? listReader.GetCount() : 0;
            if (count == 0)
                v = Array.Empty<int>();
            else
            {
                v = new int[count];
                for (int i = 0; i < count; i++)
                {
                    int item = default;
                    _Read(listReader, ref item);                    
                    v[i] = item;
                }
            }
        }
		#endregion
	}
}
